Final/multinblkserver.cpp

```diff
@@ -0,0 +1,129 @@
+	if (bound < 0)
+		error("ERROR binding to socket");
+
+	// Step 3: Listen
+	int listening = listen(sockfd, BACKLOG_SIZE); // The socket will actively listen for incoming connections and queue them up
+	if (listening < 0)
+		error("ERROR listener failed");
+
+	fd_set read_fds;
+	struct node *head = createFDList(0); // Create empty fdlist (file descriptor list)
+	while (1) {
+		select(sockfd, &read_fds, head); // select incoming connections and handle them as necessary
+	}
+
+	// Close the connection
+	printf("Closing the connection");
+    close(sockfd);
+	return 0;
+}
+
+//Pre: sockfd is the socket bound to this server and head is a pointer to the beginning of an fdlist
+//Post: sleeps until there is an available connection and then determines if it must be accepted and read from
+void select(int sockfd, fd_set *read_fds_ref, struct node *head) {
+	printf("%s\n", "Attempting to Select . . .");
+	makeFDS(head, read_fds_ref); // make fd_set from fdlist
+	FD_SET(sockfd, read_fds_ref); // (re)add sockfd to allow for new connections
+	int maxsockfd = maxfd(sockfd, head); // determine the maximum value of file descriptor for select
+	int selected = select(maxsockfd, read_fds_ref, NULL, NULL, NULL); // sleep until there has been in a change in one of the file descriptors in the fd_set
+	if (selected >= 0) {
+		printf("%s\n", "Connection Selected");
+
+		if (!wasEchoed(read_fds_ref, head)) { // determine if the connection should be echoed
+			int remotefd = accept(sockfd); // accept connection to remote client through remotefd
+			pushToList(head, remotefd); // add remotefd to list
+		}
+	}
+	else {error("ERROR Selection failed");}
+}
+
+//Pre: sockfd is the socket bound to this server
+//Post: accept connection and return the file descriptor of connected client
+int accept(int sockfd) {
+	// Step 4: Accecpt
+	struct sockaddr *remote_host; // a pointer to a sockaddr struct that will be filled in with the address information of the remote host
+	socklen_t *remote_len; // the size of the sockaddr struct in bytes
+	printf("%s\n", "Attempting to Accept . . .");
+	int remotefd = accept(sockfd, remote_host, remote_len); // dequeue from sockfd and accept connection through remotefd
+	if (remotefd < 0)
+		error("ERROR receiving request");
+	printf("%s\n", "Connection Accecpted");
+	return remotefd;
+}
+
+//Pre: head is a pointer to the beginning of an fdlist
+//Post: return 1 if data was echoed and echoes it or return 0 if data does not need to be echoed
+bool wasEchoed(fd_set *read_fds_ref, struct node *head) {
+	struct node *cur_node = head;
+	while (cur_node != NULL) { // iterate through fdlist
+		int fd = cur_node->data;
+		if (FD_ISSET(fd, read_fds_ref)) { // check if file descriptor is ready to be read from
+			printf("%s\n", "Echoing Data");
+			removeFromList(head, fd); // close connection to client part 1
+			FD_CLR(fd, read_fds_ref); // close connection to client part 2
+
+			ThreadPool thread_pool(4); // create threadpool of size 4
```

Comment:
hmmm so unfortunately this won't actually parallelize the connection handling. you are recreating a new thread pool every single time you echo a connection (since you're declaring the thread pool here). after creating the thread pool, you ask the thread pool to spawn a single thread to echo the connection, and then you destroy the thread pool (implicitly, because the thread pool goes out of scope once the method returns). as a result, multiple connections arent actually sharing the same threadpool, and because thread pool destruction is actually a blocking operation, your program will actually block until the thread spawned by the thread pool has finished before allowing this `wasEchoed` method to return, meaning that your program currently handles connections sequentially still.

Reply:
Can I just make thread_pool a global variable to fix his?